sstables/trie/trie_writer.hh

```diff
@@ -0,0 +1,1192 @@
+// which is very sad.
+//
+// Maybe the right thing is to do is to have logs that are compiled out in production builds
+// and only enabled by developers who are actively working on the relevant feature.
+// This way we free to add as many logs as we please, without worrying at all about the performance cost.
+template <typename... Args>
+void expensive_log(seastar::logger::format_info_t<Args...> fmt, Args&&... args) {
+    if constexpr (developer_build) {
+        trie_logger.trace(std::move(fmt), std::forward<Args>(args)...);
+    }
+}
+inline void expensive_assert(bool expr, std::source_location srcloc = std::source_location::current()) {
+    if constexpr (developer_build) {
+        if (!expr) {
+            __assert_fail("", srcloc.file_name(), srcloc.line(), srcloc.function_name());
+        }
+    }
+}
+
+// We aleady have bytes_view, so perhaps it should be used here.
+// But std::span<const std::byte> is, in some sense, the standard type for this purpose.
+// std::as_bytes() exists, after all.
+using const_bytes = std::span<const std::byte>;
+
+inline constexpr uint64_t round_down(uint64_t a, uint64_t factor) {
+    return a - a % factor;
+}
+
+inline constexpr uint64_t round_up(uint64_t a, uint64_t factor) {
+    return round_down(a + factor - 1, factor);
+}
+
+// Special-purpose stack-like allocator.
+//
+// It has two allowed operations:
+// 1. Append a new buffer to the "stack".
+// 2. Shrink the stack to the given address, invalidating everything allocated after it.
+//    (In particular, it's legal to shrink the stack to the middle of some allocation,
+//    invalidating the allocation's tail but preserving its head.)
+//
+// This exists because the trie writer's allocation pattern is quite stack-like,
+// and it would be a shame not to try to take advantage of it.
+//
+// That said, this isn't really a critical optimization.
+// Last time I checked, it only saves around ~25% of work.
+// I'm not sure if it's worth the added complexity.
+class bump_allocator {
+public:
+    // This is just a sanitizing wrapper around a pointer.
+    // In release builds it should behave exactly like a regular pointer.
+    template <typename T>
+    struct ptr {
+        using value_type = std::conditional_t<std::is_array_v<T>, std::remove_extent_t<T>, T>;
+        friend bump_allocator;
+        value_type* _ptr = nullptr;
+    #if TRIE_SANITIZE_BUMP_ALLOCATOR
+        // Owner.
+        bump_allocator* _alctr = nullptr;
+        // The index of last discard at the time this allocation was created.
+        size_t _discard_counter = 0;
+        // The position of this allocation in the allocator's stack.
+        size_t _global_pos = 0;
+        // The size of this allocation, in bytes.
+        // Only a past-the-end pointer, obtained with x.offset(x._size), might have _size equal to zero.
+        // Other pointers must have non-zero _size.
+        size_t _size = 0;
+    #endif
```

Comment:
Can we integrate this with asan? There are hooks to poison/unpoison memory, and we can leave poisoned gaps in debug mode. Then pointer debugging will be done by asan.